targets/XPlatCppSdk/source/build/Android/XPlatCppSdk/src/main/java/com/microsoft/xplatcppsdk/HttpRequest.java

```diff
@@ -0,0 +1,141 @@
+    private String method = "POST";
+
+    static {
+        httpClient = new OkHttpClient.Builder()
+                .retryOnConnectionFailure(false) // Explicitly disable retries;
+                .connectTimeout(15, TimeUnit.SECONDS)
+                .writeTimeout(15, TimeUnit.SECONDS)
+                .readTimeout(15, TimeUnit.SECONDS)
+                .build();
+    }
+
+    public HttpRequest() {
+        this.requestBuilder = new Request.Builder();
+    }
+
+    public void setMethod(String method) {
+        this.method = method;
+    }
+
+    public boolean setUrl(String url) {
+        try
+        {
+            this.requestBuilder = this.requestBuilder.url(url);
+        }
+        catch (Exception e)
+        {
+            return false;
+        }
+        return true;
+    }
+
+    public void setHeader(String name, String value) {
+        try
+        {
+            if(name.equalsIgnoreCase("Content-Type"))
+            {
+                this.contentType = MediaType.parse(value);
+            }
+            else
+            {
+                this.requestBuilder = requestBuilder.addHeader(name, value);
+            }
+        }
+        catch (Exception e)
+        {
+
+        }
+    }
+
+    public void setBody(byte[] body) {
+        try
+        {
+            this.requestBuilder = this.requestBuilder.method(this.method, RequestBody.create(this.contentType, body));
+        }
+        catch (Exception e)
+        {
+
+        }
+    }
+
+    public boolean isRequestCompleted() {
+        return this.requestCompleted;
+    }
+
+    public boolean sendRequest() {
+        try
+        {
+            this.requestCompleted = false;
+            this.httpClient.newCall(this.requestBuilder.build()).enqueue(new Callback() {
+                @Override
+                public void onFailure(final Call call, IOException e) {
+                    onRequestFailed(e.getClass().getCanonicalName());
+                }
+
+                @Override
+                public void onResponse(Call call, final Response response) throws IOException {
+                    onRequestCompleted(response);
+                }
+            });
+        }
+        catch (Exception e)
+        {
+            return false;
+        }
+        return true;
+    }
+
+    public void onRequestFailed(String errorMessage) {
+        this.requestCompleted = true;
+    }
+
+    public void onRequestCompleted(Response httpresponse) {
+        this.requestCompleted = true;
+        this.response = httpresponse;
+    }
+
+    public int getResponseHttpCode()
+    {
+        if (this.response != null && this.requestCompleted) {
+            return this.response.code();
+        } else {
+            return 400; // Bad request
+        }
+    }
+
+    public byte[] getResponseHttpBody()
+    {
+        if (this.response != null && this.requestCompleted) {
+            try {
+                return this.response.body().bytes();
+            } catch (IOException e) {
```

Comment:
>(IOException e) { [](start = 20, length = 17)

Do you also need to catch a generic expection? We should not let a http stack throw an unhandled exception right since we're communicating read failure via null return values ?

Reply:
This is due to the this method called from JNI. And the exception from JAVA didn't forward to JNI. And after got exception form Java side, all JNI calling is failed. And the JNI is implemented base on C, so there is no exception on it. So we need to just return error with null. And there is only one exception which is throw is "IOException" in bytes().

---
In reply to: [274664851](https://github.com/PlayFab/SDKGenerator/pull/470#discussion_r274664851) [](ancestors = 274664851)